src/AsyncInitializationWithSimpleInjectorDemo/Program.cs

```diff
@@ -41,29 +42,34 @@ internal static class AsyncInitializationExtensions
 		/// <returns></returns>
 		public static async Task<IWebHost> InitializeAsync(this IWebHost host)
 		{
-			using (var scope = host.Services.CreateScope())
-			{
-				// register any infrastructure components that have already been registered with the .NET Core DI container
-				// you only need to register the infrastructure components that are required by asynchronous initializers
-				var container = new Container();
-				container.RegisterInstance(scope.ServiceProvider.GetService<DbContextOptions<SchoolContext>>());
-
-				// register the individual async initializers in the order you want to execute them
-				container.Collection.Register<IAsyncInitializer>(
-					typeof(EntityFrameworkDatabaseMigrationAsyncInitializer),
-					typeof(EntityFrameworkDatabaseSeedDataAsyncInitializer));
-				container.Register<RootInitializer>();
-
-				// register all your application components
-				container.RegisterApplicationComponentsAndVerify();
-
-				// execute all asynchronous initializers specified above
-				await container.GetInstance<RootInitializer>().InitializeAsync(CancellationToken.None);
-			}
+			await host.Services.BuildInitializationContainerAndPerformInitialization();
 
 			return host;
 		}
 
 		public static async Task RunAsync(this Task<IWebHost> host) => await (await host).RunAsync();
+
+		// used by InitializeAsync above and for setup in integration tests
+		public static async Task BuildInitializationContainerAndPerformInitialization(this IServiceProvider serviceProvider)
```

Comment:
Moved asynchronous initialization code into its own extension method off `IServiceProvider` so that code could be shared between production and integration test setup